artemis-protocols/artemis-amqp-protocol/src/main/java/org/apache/activemq/artemis/protocol/amqp/proton/ProtonServerReceiverContext.java

```diff
@@ -278,39 +283,66 @@ public void onMessage(Delivery delivery) throws ActiveMQAMQPException {
          return;
       }
 
-      if (delivery.isAborted()) {
-         // Aborting implicitly remotely settles, so advance
-         // receiver to the next delivery and settle locally.
-         receiver.advance();
-         delivery.settle();
+      try {
+         if (delivery.isAborted()) {
+            // Aborting implicitly remotely settles, so advance
+            // receiver to the next delivery and settle locally.
+            receiver.advance();
+            delivery.settle();
+
+            // Replenish the credit if not doing a drain
+            if (!receiver.getDrain()) {
+               receiver.flow(1);
+            }
+
+            return;
+         } else if (delivery.isPartial()) {
```

Comment:
I think you need to add some more metrics here on what constitutes a "large" message as there isn't anything that requires a client to strictly adhere to max frame size limits other that it shouldn't go over, it can certainly send under that amount.  A client could split a 4k message into 4 1k chunks etc so assuming that a message is large just based on partial will lead to likely considering messages that don't meet the normal Artemis definition of a large message as being large.  Also even if not partial you ought to be checking if what you have in a complete message that meets your definition of a large message as the AMQP max frame size is configurable on the protocol adapter so you might need a second metric there to say it's not partial but it exceeds what we consider "not large" and would prefer to store it in the large message way.

Reply:
@tabish121 makes sense.. thanks a lot.. will add another commit to the REVIEW

Reply:
I was actually looking for that information.. if it made sense to use frame size or not.

Reply:
@tabish121 commit added

These commits will be squahsed after we are done with the review.